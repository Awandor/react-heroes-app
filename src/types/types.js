export const types = {

  // login: 'login' // podría ser pero es mejor ser más específicos
  login: '[auth] login',
  logout: '[auth] logout'

};
