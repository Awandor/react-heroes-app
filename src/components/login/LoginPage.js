import { useContext } from "react";
import { useHistory } from "react-router-dom";
import { AuthContext } from "../../authentication/AuthContext";
import { types } from "../../types/types";

export const LoginPage = () => {

    const history = useHistory();

    const context = useContext( AuthContext );

    // console.log( 'context', context );

    const handleLogin = ( e ) => {

        // history.push('/marvel'); // Nos empuja a la página
        // history.replace('/marvel'); // Nos lleva a la página y reemplaza LoginPage por ésta en el historial

        const lastPath =localStorage.getItem( 'lastPath' ) || '/';

        context.dispatch( {
            type: types.login,
            payload: {
                name: 'Dan'
            }
        } );

        // history.replace( '/' ); 

        history.replace( lastPath ); 

    };

    return (
        <div className="container">
            <h1>Login</h1>
            <hr />

            <button className="btn btn-primary" onClick={handleLogin}>Login</button>
        </div>
    );

};
