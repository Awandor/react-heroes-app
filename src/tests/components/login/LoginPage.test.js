import { mount } from "enzyme";
import { MemoryRouter, Router } from "react-router-dom";
import { AuthContext } from "../../../authentication/AuthContext";
import { LoginPage } from "../../../components/login/LoginPage";
import { types } from "../../../types/types";

describe( 'Pruebas en <LoginPage />', () => {

    const contextValue = {
        dispatch: jest.fn(),
        userState: {
            logged: false
        }
    };

    const historyMock = {
        push: jest.fn(),
        location: {},

        // Nos fijamos en los errores de lo que falta y los vamos añadiendo deduciendo el tipo
        listen: jest.fn(),
        createHref: jest.fn(),
        replace: jest.fn(),
    };

    test( 'debe mostrarse correctamente', () => {

        const wrapper = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter>
                    <LoginPage />
                </MemoryRouter>
            </AuthContext.Provider>
        );

        // console.log( wrapper );

        expect( wrapper ).toMatchSnapshot();

    } );

    test( 'debe realizar el dispatch y la navegación', () => {

        const wrapperWithHistory = mount(
            <AuthContext.Provider value={contextValue}>
                <MemoryRouter>
                    <Router history={historyMock}>
                        <LoginPage />
                    </Router>
                </MemoryRouter>
            </AuthContext.Provider>
        );

        wrapperWithHistory.find( 'button' ).simulate( 'click' );

        // Vamos a evaluar si la función dispatch dentro de handleLogout ha sido invocada

        expect( contextValue.dispatch ).toHaveBeenCalled();

        // Ahora enviamos argumento y falla si no proveemos al wrapper de un history

        const userData = {
            type: types.login,
            payload: {
                name: 'Dan'
            }
        };

        expect( contextValue.dispatch ).toHaveBeenCalledWith( userData );

        // handleLogin usa history.replace, vamos a ver si se ha usado

        expect( historyMock.replace ).toHaveBeenCalledWith( '/' );

        // Vamos a usar localStorage

        localStorage.setItem( 'lastPath', '/dc' );

        wrapperWithHistory.find( 'button' ).simulate( 'click' );

        expect( historyMock.replace ).toHaveBeenCalledWith( '/dc' );

    } );

} );
