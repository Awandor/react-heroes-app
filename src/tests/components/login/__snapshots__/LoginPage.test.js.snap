// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <LoginPage /> debe mostrarse correctamente 1`] = `
<div
  className="container"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Login
  </button>
</div>
`;
